File: experimental/teams/js/adaptiveCards/src/adaptiveCardsBot.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import {
    CardFactory,
    InvokeResponse,
    MessageFactory,
    TaskModuleMessageResponse,
    TaskModuleRequest,
    TaskModuleResponseBase,
    TaskModuleTaskInfo,
    TeamsActivityHandler,
    TurnContext
} from 'botbuilder';

/**
 * You can @mention the bot the text "1", "2", or "3". "1" will send back adaptive cards. "2" will send back a
 * task module that contains an adpative card. "3" will return an adpative card that contains BF card actions.
 */
export class AdaptiveCardsBot extends TeamsActivityHandler {
    constructor() {
        super();

        // See https://aka.ms/about-bot-activity-message to learn more about the message and other activity types.
        this.onMessage(async (context, next) => {
            TurnContext.removeRecipientMention(context.activity);
            let text = context.activity.text;
            if (text && text.length > 0) {
                text = text.trim();
                switch (text) {
                    case '1':
                        await this.sendAdaptiveCard1(context);
                        break;

                    case '2':
                        await this.sendAdaptiveCard2(context);
                        break;

                    case '3':
                        await this.sendAdaptiveCard3(context);
                        break;

                    default:
                        await context.sendActivity(`You said: ${text}`);
                }
            } else {
                await context.sendActivity('App sent a message with empty text');
                const activityValue = context.activity.value;
                if (activityValue) {
                    await context.sendActivity(`but with value ${JSON.stringify(activityValue)}`);
                }
            }
            await next();
        });

        this.onMembersAdded(async (context, next) => {
            const membersAdded = context.activity.membersAdded;
            for (const member of membersAdded) {
                if (member.id !== context.activity.recipient.id) {
                    await context.sendActivity('Hello and welcome!');
                }
            }

            // By calling next() you ensure that the next BotHandler is run.
            await next();
        });
    }

    protected async onTeamsTaskModuleFetch(context: TurnContext, taskModuleRequest: TaskModuleRequest): Promise<TaskModuleTaskInfo> {
        await context.sendActivity(MessageFactory.text(`OnTeamsTaskModuleFetchAsync TaskModuleRequest: ${JSON.stringify(taskModuleRequest)}`));

        /**
         * The following line disables the lint rules for the Adaptive Card so that users can
         * easily copy-paste the object into the Adaptive Cards Designer to modify their cards
         * The Designer can be found here: https://adaptivecards.io/designer/
         */
        /* tslint:disable:quotemark object-literal-key-quotes */
        const card = CardFactory.adaptiveCard({
            "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
            "actions": [
                {
                    "data": {
                        "submitLocation": "taskModule"
                    },
                    "title": "Action.Submit",
                    "type": "Action.Submit"
                }
            ],
            "body": [
                {
                    "text": "This is an Adaptive Card within a Task Module",
                    "type": "TextBlock"
                }
            ],
            "type": "AdaptiveCard",
            "version": "1.0"
        });
        /* tslint:enable:quotemark object-literal-key-quotes */
        return {
            card,
            height: 200,
            title: 'Task Module Example',
            width: 400
        } as TaskModuleTaskInfo;
    }

    protected async onTeamsTaskModuleSubmit(context: TurnContext, taskModuleRequest: TaskModuleRequest): Promise<TaskModuleResponseBase> {
        await context.sendActivity(MessageFactory.text(`OnTeamsTaskModuleSubmit value: ${JSON.stringify(taskModuleRequest)}`));
        return { type: 'message', value: 'Thanks!' } as TaskModuleMessageResponse;
    }

    protected async onTeamsCardActionInvoke(context: TurnContext): Promise<InvokeResponse> {
        await context.sendActivity(MessageFactory.text(`OnTeamsCardActionInvoke value: ${JSON.stringify(context.activity.value)}`));
        return { status: 200 } as InvokeResponse;
    }

    private async sendAdaptiveCard1(context: TurnContext): Promise<void> {
        /* tslint:disable:quotemark object-literal-key-quotes */
        const card = CardFactory.adaptiveCard({
            "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
            "actions": [
                {
                    "data": {
                        "msteams": {
                            "type": "imBack",
                            "value": "text"
                        }
                    },
                    "title": "imBack",
                    "type": "Action.Submit"
                },
                {
                    "data": {
                        "msteams": {
                            "type": "messageBack",
                            "value": { "key": "value" }
                        }
                    },
                    "title": "message back",
                    "type": "Action.Submit"
                },
                {
                    "data": {
                        "msteams": {
                            "displayText": "display text message back",
                            "text": "text received by bots",
                            "type": "messageBack",
                            "value": { "key": "value" }
                        }
                    },
                    "title": "message back local echo",
                    "type": "Action.Submit"
                },
                {
                    "data": {
                        "msteams": {
                            "type": "invoke",
                            "value": { "key": "value" }
                        }
                    },
                    "title": "invoke",
                    "type": "Action.Submit"
                }
            ],
            "body": [
                {
                    "text": "Bot Builder actions",
                    "type": "TextBlock"
                }
            ],
            "type": "AdaptiveCard",
            "version": "1.0"
        });
        /* tslint:enable:quotemark object-literal-key-quotes */
        await context.sendActivity(MessageFactory.attachment(card));
    }

    private async sendAdaptiveCard2(context: TurnContext): Promise<void> {
        /* tslint:disable:quotemark object-literal-key-quotes */
        const card = CardFactory.adaptiveCard({
            "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
            "actions": [
                {
                    "data": {
                        "msteams": {
                            "type": "invoke",
                            "value": {
                                "hiddenKey": "hidden value from task module launcher",
                                "type": "task/fetch"
                            }
                        }
                    },
                    "title": "Launch Task Module",
                    "type": "Action.Submit"
                }
            ],
            "body": [
                {
                    "text": "Task Module Adaptive Card",
                    "type": "TextBlock"
                }
            ],
            "type": "AdaptiveCard",
            "version": "1.0"
        });
        /* tslint:enable:quotemark object-literal-key-quotes */
        await context.sendActivity(MessageFactory.attachment(card));
    }

    private async sendAdaptiveCard3(context: TurnContext): Promise<void> {
        /* tslint:disable:quotemark object-literal-key-quotes */
        const card = CardFactory.adaptiveCard({
            "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
            "actions": [
                {
                    "data": {
                        "key": "value"
                    },
                    "title": "Action.Submit",
                    "type": "Action.Submit"
                }
            ],
            "body": [
                {
                    "text": "Bot Builder actions",
                    "type": "TextBlock"
                },
                {
                    "id": "x",
                    "type": "Input.Text"
                }
            ],
            "type": "AdaptiveCard",
            "version": "1.0"
        });
        /* tslint:enable:quotemark object-literal-key-quotes */
        await context.sendActivity(MessageFactory.attachment(card));
    }
}

File: experimental/teams/csharp-dotnetcore/Teams/AdaptiveCards/Bots/AdaptiveCardsBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using AdaptiveCards;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Teams;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Microsoft.BotBuilderSamples.Bots
{
    public class AdaptiveCardsBot : TeamsActivityHandler
    {
        /*
         * You can @mention the bot the text "1", "2", or "3". "1" will send back adaptive cards. "2" will send back a 
         * task module that contains an adpative card. "3" will return an adpative card that contains BF card actions.
         * 
         */
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {

            if (turnContext.Activity.Text != null)
            {
                turnContext.Activity.RemoveRecipientMention();
                string actualText = turnContext.Activity.Text;

                if (actualText.Equals("1"))
                {
                    var adaptiveCard = new AdaptiveCard();
                    adaptiveCard.Body.Add(new AdaptiveTextBlock("Bot Builder actions"));

                    var action1 = new CardAction(ActionTypes.ImBack, "imBack", null, null, null, "text");
                    var action2 = new CardAction(ActionTypes.MessageBack, "message back", null, null, null, JObject.Parse(@"{ ""key"" : ""value"" }"));
                    var action3 = new CardAction(ActionTypes.MessageBack, "message back local echo", null, "text received by bots", "display text message back", JObject.Parse(@"{ ""key"" : ""value"" }"));
                    var action4 = new CardAction("invoke", "invoke", null, null, null, JObject.Parse(@"{ ""key"" : ""value"" }"));

                    adaptiveCard.Actions.Add(action1.ToAdaptiveCardAction());
                    adaptiveCard.Actions.Add(action2.ToAdaptiveCardAction());
                    adaptiveCard.Actions.Add(action3.ToAdaptiveCardAction());
                    adaptiveCard.Actions.Add(action4.ToAdaptiveCardAction());

                    var replyActivity = MessageFactory.Attachment(adaptiveCard.ToAttachment());
                    await turnContext.SendActivityAsync(replyActivity, cancellationToken);
                }
                else if (actualText.Equals("2"))
                {
                    var taskModuleAction = new TaskModuleAction("Launch Task Module", @"{ ""hiddenKey"": ""hidden value from task module launcher"" }");

                    var adaptiveCard = new AdaptiveCard();
                    adaptiveCard.Body.Add(new AdaptiveTextBlock("Task Module Adaptive Card"));
                    adaptiveCard.Actions.Add(taskModuleAction.ToAdaptiveCardAction());

                    var replyActivity = MessageFactory.Attachment(adaptiveCard.ToAttachment());
                    await turnContext.SendActivityAsync(replyActivity, cancellationToken);
                }
                else if (actualText.Equals("3"))
                {
                    var adaptiveCard = new AdaptiveCard();
                    adaptiveCard.Body.Add(new AdaptiveTextBlock("Bot Builder actions"));
                    adaptiveCard.Body.Add(new AdaptiveTextInput { Id = "x" });
                    adaptiveCard.Actions.Add(new AdaptiveSubmitAction { Type = "Action.Submit", Title = "Action.Submit", Data = new JObject { { "key", "value" } } });

                    var replyActivity = MessageFactory.Attachment(adaptiveCard.ToAttachment());
                    await turnContext.SendActivityAsync(replyActivity, cancellationToken);
                }
                else
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text($"You said: {turnContext.Activity.Text}"), cancellationToken);
                }
            }
            else
            {
                await turnContext.SendActivityAsync(MessageFactory.Text($"App sent a message with null text"), cancellationToken);

                // TODO: process the response from the card

                if (turnContext.Activity.Value != null)
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text($"but with value {turnContext.Activity.Value}"), cancellationToken);
                }
            }
        }

        protected override async Task<TaskModuleTaskInfo> OnTeamsTaskModuleFetchAsync(ITurnContext<IInvokeActivity> turnContext, TaskModuleRequest taskModuleRequest, CancellationToken cancellationToken)
        {
            var reply = MessageFactory.Text("OnTeamsTaskModuleFetchAsync TaskModuleRequest: " + JsonConvert.SerializeObject(taskModuleRequest));
            await turnContext.SendActivityAsync(reply, cancellationToken);

            var adaptiveCard = new AdaptiveCard();
            adaptiveCard.Body.Add(new AdaptiveTextBlock("This is an Adaptive Card within a Task Module"));
            adaptiveCard.Actions.Add(new AdaptiveSubmitAction { Type = "Action.Submit", Title = "Action.Submit", Data = new JObject { { "submitLocation", "taskModule" } } });

            return new TaskModuleTaskInfo()
            {
                Card = adaptiveCard.ToAttachment(),
                Height = 200,
                Width = 400,
                Title = "Task Module Example",
            };
        }

        protected override async Task<TaskModuleResponseBase> OnTeamsTaskModuleSubmitAsync(ITurnContext<IInvokeActivity> turnContext, TaskModuleRequest taskModuleRequest, CancellationToken cancellationToken)
        {
            await turnContext.SendActivityAsync(MessageFactory.Text($"OnTeamsTaskModuleSubmitAsync value: { JsonConvert.SerializeObject(taskModuleRequest) }"), cancellationToken);
            return new TaskModuleMessageResponse { Value = "Thanks!" };
        }

        protected override async Task<InvokeResponse> OnTeamsCardActionInvokeAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
        {
            await turnContext.SendActivityAsync(MessageFactory.Text($"OnTeamsCardActionInvokeAsync value: {turnContext.Activity.Value}"), cancellationToken);
            return new InvokeResponse() { Status = 200 };
        }
    }
}
